hack/test-go.sh

```diff
@@ -1,125 +1,231 @@
-}
+# TODO(skuznets): refactor after https://github.com/openshift/origin/pull/6591
+BASETMPDIR="${TMPDIR:-/tmp}/openshift/test-go"
+LOG_DIR="${BASETMPDIR}/logs"
+ARTIFACT_DIR="${BASETMPDIR}/artifacts"
+mkdir -p "${BASETMPDIR}" "${LOG_DIR}" "${ARTIFACT_DIR}"
 
-special_upstream_test_dirs() {
-  echo "${OS_GO_PACKAGE}/${KUBE_GODEP_PATH}/api"
-  echo "${OS_GO_PACKAGE}/${KUBE_GODEP_PATH}/api/v1"
-}
+# TODO(skuznets): remove these once we've migrated all tools to the new vars
+if [[ -n "${KUBE_TIMEOUT+x}" ]]; then
+    TIMEOUT="${KUBE_TIMEOUT}"
+    echo "[WARNING] The flag \$KUBE_TIMEOUT for $0 is deprecated, use \$TIMEOUT instead."
+fi
 
-# find the upstream test directories, excluding special-case directories and the upstream runtime package.
-# The tests for the upstream runtime package are not solvent currently due to a patch for:
-# https://github.com/kubernetes/kubernetes/pull/9971
-find_upstream_test_dirs() {
-  cd "${OS_ROOT}"
-  find ./Godeps/_workspace/src/k8s.io/kubernetes -not \( \
-      \( \
-        -wholename "${KUBE_GODEP_PATH}/api" \
-        -o -wholename "${KUBE_GODEP_PATH}/api/v1" \
-        -o -wholename './Godeps/_workspace/src/k8s.io/kubernetes/pkg/runtime' \
-        -o -wholename './Godeps/_workspace/src/k8s.io/kubernetes/test/e2e' \
-      \) -prune \
-    \) -name '*_test.go' -print0 | xargs -0n1 dirname | sort -u | xargs -n1 printf "${OS_GO_PACKAGE}/%s\n"
-}
+if [[ -n "${KUBE_COVER+x}" ]]; then
+    COVERAGE_SPEC="${KUBE_COVER}"
+    echo "[WARNING] The flag \$KUBE_COVER for $0 is deprecated, use \$COVERAGE_SPEC instead."
+fi
 
-# -covermode=atomic becomes default with -race in Go >=1.3
-if [ -z ${KUBE_COVER+x} ]; then
-  KUBE_COVER=""
+if [[ -n "${OUTPUT_COVERAGE+x}" ]]; then
+    COVERAGE_OUTPUT_DIR="${OUTPUT_COVERAGE}"
+    echo "[WARNING] The flag \$OUTPUT_COVERAGE for $0 is deprecated, use \$COVERAGE_OUTPUT_DIR instead."
 fi
 
-OUTPUT_COVERAGE=${OUTPUT_COVERAGE:-""}
+if [[ -n "${KUBE_RACE+x}" ]]; then
+    DETECT_RACES="${KUBE_RACE}"
+    echo "[WARNING] The flag \$KUBE_RACE for $0 is deprecated, use \$DETECT_RACES instead."
+fi
 
-if [ -n "${OUTPUT_COVERAGE}" ]; then
-  if [ -z ${KUBE_RACE+x} ]; then
-    KUBE_RACE="-race"
-  fi
-  if [ -z "${KUBE_COVER}" ]; then
-    KUBE_COVER="-cover -covermode=atomic"
-  fi
+# Internalize envionment variables we consume and default if they're not set
+print_packages="${PRINT_PACKAGES:-}"
+test_kube="${TEST_KUBE:-}"
+test_timeout="${TIMEOUT:-60s}"
+detect_races="${DETECT_RACES:-true}"
+coverage_output_dir="${COVERAGE_OUTPUT_DIR:-}"
+coverage_spec="${COVERAGE_SPEC:--cover -covermode atomic}"
+gotest_flags="${GOTEST_FLAGS:-}"
+junit_report="${JUNIT_REPORT:-}"
+
+if [[ -n "${junit_report}" && -n "${coverage_output_dir}" ]]; then
+    echo "$0 cannot create jUnit XML reports and coverage reports at the same time."
+    exit 1
 fi
 
-if [ -z ${KUBE_RACE+x} ]; then
-  KUBE_RACE=""
+# determine if user wanted verbosity
+verbose=
+if echo "${gotest_flags}" | grep -Eq "\-v( |$)"; then
```

Comment:
It would be simpler to use `if  [[ "${gotest_flags}" =~ -v( |$) ]]`.  If you prefer using `grep`, I'd suggest using `-e` before the pattern instead of prepending the pattern with a backslash, and you can use `\>` instead of `( |$)`.